core/lib/types/src/abi.rs

```diff
@@ -0,0 +1,343 @@
+        ])
+    }
+
+    pub fn decode(token: Token) -> anyhow::Result<Self> {
+        let tokens = token.into_tuple().context("not a tuple")?;
+        anyhow::ensure!(tokens.len() == 16);
+        let mut t = tokens.into_iter();
+        let mut next = || t.next().unwrap();
+        Ok(Self {
+            tx_type: next().into_uint().context("tx_type")?,
+            from: next().into_uint().context("from")?,
+            to: next().into_uint().context("to")?,
+            gas_limit: next().into_uint().context("gas_limit")?,
+            gas_per_pubdata_byte_limit: next().into_uint().context("gas_per_pubdata_byte_limit")?,
+            max_fee_per_gas: next().into_uint().context("max_fee_per_gas")?,
+            max_priority_fee_per_gas: next().into_uint().context("max_priority_fee_per_gas")?,
+            paymaster: next().into_uint().context("paymaster")?,
+            nonce: next().into_uint().context("nonce")?,
+            value: next().into_uint().context("value")?,
+            reserved: next()
+                .into_fixed_array()
+                .context("reserved")?
+                .into_iter()
+                .enumerate()
+                .map(|(i, t)| t.into_uint().context(i))
+                .collect::<Result<Vec<_>, _>>()
+                .context("reserved")?
+                .try_into()
+                .ok()
+                .context("reserved")?,
+            data: next().into_bytes().context("data")?,
+            signature: next().into_bytes().context("signature")?,
+            factory_deps: next()
+                .into_array()
+                .context("factory_deps")?
+                .into_iter()
+                .enumerate()
+                .map(|(i, t)| t.into_uint().context(i))
+                .collect::<Result<_, _>>()
+                .context("factory_deps")?,
+            paymaster_input: next().into_bytes().context("paymaster_input")?,
+            reserved_dynamic: next().into_bytes().context("reserved_dynamic")?,
+        })
+    }
+
+    pub fn encode(&self) -> Token {
+        Token::Tuple(vec![
+            Token::Uint(self.tx_type),
+            Token::Uint(self.from),
+            Token::Uint(self.to),
+            Token::Uint(self.gas_limit),
+            Token::Uint(self.gas_per_pubdata_byte_limit),
+            Token::Uint(self.max_fee_per_gas),
+            Token::Uint(self.max_priority_fee_per_gas),
+            Token::Uint(self.paymaster),
+            Token::Uint(self.nonce),
+            Token::Uint(self.value),
+            Token::FixedArray(self.reserved.iter().map(|x| Token::Uint(*x)).collect()),
+            Token::Bytes(self.data.clone()),
+            Token::Bytes(self.signature.clone()),
+            Token::Array(self.factory_deps.iter().map(|x| Token::Uint(*x)).collect()),
+            Token::Bytes(self.paymaster_input.clone()),
+            Token::Bytes(self.reserved_dynamic.clone()),
+        ])
+    }
+
+    pub fn hash(&self) -> H256 {
```

Comment:
just in case, this hash is correct only for L1->L2 transactions. The rest have special functions to calculate the correct hash of the transaction, so better to at least add a comment about it